base/regex.jl

```diff
@@ -282,22 +279,20 @@ function matchall(re::Regex, str::String, overlap::Bool=false)
     matches
 end
 
-matchall(re::Regex, str::SubString, overlap::Bool=false) =
+matchall(re::Regex, str::AbstractString, overlap::Bool=false) =
     matchall(re, String(str), overlap)
 
-function search(str::Union{String,SubString}, re::Regex, idx::Integer)
-    if idx > nextind(str,endof(str))
-        throw(BoundsError())
-    end
+function search(str::Union{String,SubString{String}}, re::Regex, idx::Integer)
+    isvalid(str, idx) || throw(ArgumentError("invalid index $i"))
```

Comment:
"invalid character index" would be more explicit. Also throw a `UnicodeError` for consistency?

Reply:
OK. I will separate this into `BoundsError` and `UnicodeError`. It will introduce a slight overhead though as `isvalid` does not distinguish them.

Reply:
Ah, that's annoying. Make it throw either `BoundsError` or `UnicodeError`, and let's say we'll improve that later by replacing `isvalid` everywhere it's appropriate with a function which would throw the errors directly.

Reply:
I reverted it to a corrected version of the original test to minimize the number of changes in this PR (as this change is breaking). I propose to implement more strict testing when your search Julep is decided.